### ISSUE TITLE
`Device Not Found` Error After Device Is Idle for a Few Minutes

### ISSUE THREAD

**Issue by idogrf** [fondberg/spotcast#450]
# Bug Ticket

## Describe the bug
Hi, 

I've recently started using SpotCast again after a while.
In my setup, I have two devices I can cast Spotify to:
- Google Nest Mini 2
- Yamaha YAS-209

Both devices are connected to Home Assistant and work fine with all other automations, integrations, and playback tasks.

The issue is that if I try to cast to one of these devices after they've been idle for a few minutes, I get a Device not found error in SpotCast.
I'm using the spotcast.start service and it doesn't matter if I use `spotify_device_id`, `device_name`, or `entity_id`.

However, if I first cast to one of these devices through the Spotify app on my phone, then SpotCast works fine, and the device shows up again.

I've also checked the [Spotify Developer API](https://developer.spotify.com/documentation/web-api/reference/get-a-users-available-devices) when SpotCast can't find the devices, and the /devices response doesn't list the idle ones.

Not sure if there's something wrong with my setup, or if I'm missing something in the way I work.

<A clear and concise description of what the bug is.>

## Troubleshooting

Make sure to validate all the elements before submitting the ticket (Exception to the steps marked as optional)

* [x] Using latest version of spotcast - **3.8.0**
* [x] Using latest stable version of Home Assistant - **2024.5.1**
* [x] I have setup the Spotify integration in Home Assistant
* [x] I have renewed my `sp_dc` and `sp_key` values and restarted Home Assistant (see [README](https://github.com/fondberg/spotcast#enabling-debug-log))
* [ ] (optional) I have Spotify Premium
* [ ] (optional) I am using multiple accounts
* [ ] (optional) I'm attaching relevant logs with level debug for component spotcast (see [README](https://github.com/fondberg/spotcast#enabling-debug-log))
* [x] (optional) I'm using entity_id in the service call and have tried device_name but the issue remains

## Environment

 - Installation type: [Container]
 - HA version: [2024.5.1]
 - spotcast version: [3.8.0]

## Configuration

````yaml
# please remove any sensitive information like cookies and token keys
spotcast:
  sp_dc: "###"
  sp_key: "###"

````

## Service Call

If relevant, provide a `yaml` of the service call or explain the action taken to replicate the issue.

````yaml
service: spotcast.start
data:
  spotify_device_id: ###
  uri: spotify:playlist:###
  random_song: true
  shuffle: true
````

## Logs

* normal|debug

````log
stderr 08/13/2024 09:01:04 PM spotipy.exceptions.SpotifyException: http status: 404, code:-1 - https://api.spotify.com/v1/me/player/play?device_id=###:
stderr 08/13/2024 09:01:04 PM Device not found, reason: None
````

## Additional context

<Add any other context about the problem here.>

**Comment by fcusson** [comment]
@idogrf I can confirm Spotcast will not be able to wake your device if you use spotify_id. Spotify ID is for devices not integrated with Home Assistant, but require the device to be woken manually and be available before making the service call.

Spotcast should still be able to wake your device prior to make the play uri request. Do you here a wakeup sound from your device when you are trying to do the service call using the entity id?

**Comment by idogrf** [comment]
Hi @fcusson thanks for the reply.

So if I understand correctly, when using devices that are integrated to Home Assistant, if I use the entity id, correct?

Let's focus on the Google Nest Mini 2 which I assume is integrated to Home Assistant.
If I try to call the service like this - 
```
service: spotcast.start
data:
  uri: spotify:playlist:###
  random_song: true
  shuffle: true
  entity_id: media_player.living_room_speaker
```
I get an error saying `Could not find device Living Room Speaker from hass.data`

here's the full error - 
```
2024-08-13T21:26:39.015779651Z [31m2024-08-14 00:26:39.014 ERROR (SyncWorker_13) [custom_components.spotcast.spotcast_controller] Could not find device Living Room Speaker from hass.data[0m
2024-08-13T21:26:39.019982164Z [31m2024-08-14 00:26:39.016 ERROR (MainThread) [homeassistant.helpers.script.websocket_api_script] websocket_api script: Error executing script. Error for call_service at pos 1: Could not find device with name Living Room Speaker[0m
2024-08-13T21:26:39.021628462Z [31m2024-08-14 00:26:39.018 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [547067348800] Error handling message: Could not find device with name Living Room Speaker (home_assistant_error) Ido Greenfeld from 192.168.31.70 (Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36)[0m
```

Is something wrong with my setup?

**Comment by fcusson** [comment]
@idogrf 

> So if I understand correctly, when using devices that are integrated to Home Assistant, if I use the entity id it should wakr the device up, correct?

Or device name, but yes if its in home assistant, we can use that information to try and wake it up. The spotify id requires the device to be up already, we can,t communicate with the google chromecast api to make a wake up command.

Did you rename the device? Is it Living Room Speaker in the google home app? We are using the name of the device for when we communicate to the chromecast API. By default, the name that Home Assistant sets when creating the device is the proper one

**Comment by idogrf** [comment]
I think I got it working!

Your mention of using the `entity_id` led me to search for similar issues, and I discovered that the speaker name was different in Home Assistant compared to the Google Home app `Living Room speaker` instead of `Living Room Speaker`.

It might be worth explaining the different uses of `entity_id` vs `spotify_device_id` in the readme (maybe I missed it?)
I'll try and open a PR for that.

Thanks for the assist!

**Comment by idogrf** [comment]
Hi @fcusson, the issue is resolved for my Google Nest Mini 2, but I'm still having trouble with my Yamaha YAS-209.

When using the Spotify Device ID, it works fine as long as the device isn't idle, which makes sense based on what we've discussed.

However, when I use the entity_id, I get a homeassistant.exceptions.HomeAssistantError: Could not find device with name Living Room Soundbar error. Unlike the Nest Mini, the YAS-209 doesn't wake up from idle.

Is there any way to confirm if this device is even supported? It’s not integrated via the Google Cast integration but instead through an unofficial integration ([Yamaha Soundbar](https://github.com/osk2/yamaha-soundbar)).

I'm reopening the same issue, If you want me to open this in a new one let me know.

**Comment by fcusson** [comment]
@idogrf if it's not a Chromecast device then spotcast is only able to call it if it is registered with Spotify (by that I mean broadcasting itself in your device list

Spotcast automates the process of waking, waiting for registration and then starting playabck for Chromecast device using a combination of the Chromecast Api and the spotify Api.

If your device uses a different protocol then it is not possible (with the current capability of the integration) to wake your device.

If anyone that has a Yamaha speaker is able to test and provide a Pr I would gladly merge it, but without the physical device it would be very hard to code that myself.

I the meantime, if you have an integration working for HA you could test if there is something you can do to wake up the device like forcing a power on and then waiting. This could do the trick for you.

I the readme, we have an Api endpoint from Spotify that you can use to see the listed devices I assume that is how you found your spotify ID, you can use that to monitor for device registration.

**Comment by idogrf** [comment]
OK, got you
I'll try and see if there's a way to wake the device automatically.
Thanks for all the help 🙏

**Comment by idogrf** [comment]
For anyone dealing with a similar issue with the YAS-209 or any other non chrome cast devices:

I discovered that the device would disappear from my [list of available devices](https://developer.spotify.com/documentation/web-api/reference/get-a-users-available-devices), which prevented SpotCast from being able to cast to it. It turns out this was happening because another Spotify account was casting to the device. When that happened, the device would vanish from my list and show up in the other account's list.

What I did to fix this was to add the other account to SpotCast as well (as explained [here](https://github.com/fondberg/spotcast?tab=readme-ov-file#multiple-accounts)) and then call the SpotCast endpoint twice—once for each account. This way, one of them would connect and start playing.
